pkg/pillar/cmd/zedrouter/appcontainer.go

```diff
@@ -101,3 +112,42 @@ func appContainerGetStats(ipAddr net.IP) (types.AppContainerMetrics, error) {
 	// XXX collect container stats for each container with the docker API endpoint ipAddr
 	return acMetrics, nil
 }
+
+func appChangeContainerStatsACL(newIPAddr, oldIPAddr net.IP) {
+	if oldIPAddr != nil {
+		// remove the previous installed blocking ACL
+		appConfigContainerStatsACL(oldIPAddr, true)
+	}
+	if newIPAddr != nil {
+		// install the App Container blocking ACL
+		appConfigContainerStatsACL(newIPAddr, false)
+	}
+}
+
+func appConfigContainerStatsACL(appIPAddr net.IP, isRemove bool) {
+	var action string
+	if isRemove {
+		action = "-D"
+	} else {
+		action = "-A"
+	}
+	// install or remove the App Container Stats blocking ACL
+	// This ACL blocks the other Apps accessing through the same subnet to 'appIPAddr:DOCKERAPIPORT'
+	// in TCP protocol, and only allow the Dom0 process to query the App's docker stats
+	// - this blocking is only possible in the 'raw' table and 'PREROUTING' chain due to the marking
+	//   is done in the 'mangle' of 'PREROUTING'
+	// - this blocking ACL does not block the Dom0 access to the above TCP endpoint on the same
+	//   subnet. This is due to the IP packets from Dom0 to the internal bridge entering the linux
+	//   forwarding through the 'OUTPUT' chain
+	// - this blocking does not seem to work if further matching to the '--physdev', so the drop action
+	//   needs to be at network layer3
+	// - XXX currently the 'drop' mark of 0xffffff on the flow of internal traffic on bridge does not work,
+	//   later it may be possible to change below '-j DROP' to '-j MARK' action
+	err := iptables.IptableCmd("-t", "raw", action, "PREROUTING", "-d", appIPAddr.String(), "-p", "tcp",
```

Comment:
User configured ACL rules are added from acl.go
Does your rule always end up at the top of the rule list in PREROUTING chain?
Otherwise, the packets that you are trying to block may get accepted by user configured rules (eg: accept all rule).

Reply:
@gkodali-zededa on Azure App I have, I do have *, and 0.0.0.0/0 both on the outbound direction, and this rule is appended on my device at the end of PREROUTING. but let me try to see how to make it to the top.

Reply:
@naiming-zededa That is a bit odd.
If the user ACL with 0.0.0.0/0 match is above your drop rule, packet should match the user rule and get accepted. Not sure what is happening there.

Can you please paste a snapshot of how rules in raw table look like with your app and drop rule?

Reply:
@gkodali-zededa this was in raw table:
(ns: pillar) 10d0ffa2-8df1-4c1b-be74-a8b30b0333e1:/# iptables -S -t raw
-P PREROUTING ACCEPT
-P OUTPUT ACCEPT
-A PREROUTING -i bn1 -p udp -m physdev --physdev-in nbu1x1+ -m set --match-set ipv4.local dst -m udp --dport 67 -j ACCEPT
-A PREROUTING -i bn1 -p udp -m physdev --physdev-in nbu1x1+ -m set --match-set ipv4.local src -m udp --sport 67 -j ACCEPT
-A PREROUTING -d 10.1.0.1/32 -i bn1 -p udp -m physdev --physdev-in nbu1x1+ -m udp --dport 67 -j ACCEPT
-A PREROUTING -s 10.1.0.1/32 -i bn1 -p udp -m physdev --physdev-in nbu1x1+ -m udp --sport 67 -j ACCEPT
-A PREROUTING -d 10.1.0.1/32 -i bn1 -p udp -m physdev --physdev-in nbu1x1+ -m udp --dport 53 -j ACCEPT
-A PREROUTING -s 10.1.0.1/32 -i bn1 -p udp -m physdev --physdev-in nbu1x1+ -m udp --sport 53 -j ACCEPT
-A PREROUTING -d 10.1.0.1/32 -i bn1 -p tcp -m physdev --physdev-in nbu1x1+ -m tcp --dport 53 -j ACCEPT
-A PREROUTING -s 10.1.0.1/32 -i bn1 -p tcp -m physdev --physdev-in nbu1x1+ -m tcp --sport 53 -j ACCEPT
-A PREROUTING -i bn1 -m physdev --physdev-in nbu1x1+ -m set --match-set ipv4. dst -j ACCEPT
-A PREROUTING -i bn1 -m physdev --physdev-in nbu1x1+ -j ACCEPT
-A PREROUTING -i bn1 -p tcp -m physdev --physdev-in nbu1x1+ -m tcp --sport 22 -j ACCEPT
-A PREROUTING -i bn1 -p tcp -m physdev --physdev-in nbu1x1+ -m tcp --sport 8085 -j ACCEPT
-A PREROUTING -i bn1 -m physdev --physdev-in nbu1x1+ -j LOG --log-prefix "FORWARD:FROM:" --log-level 3

Reply:
this is in filter table:
(ns: pillar) 10d0ffa2-8df1-4c1b-be74-a8b30b0333e1:/# iptables -S
-P INPUT ACCEPT
-P FORWARD ACCEPT
-P OUTPUT ACCEPT
-A INPUT -p tcp -m tcp --dport 8080 -j REJECT --reject-with tcp-reset
-A INPUT -s 127.0.0.1/32 -d 127.0.0.1/32 -p tcp -m tcp --dport 4822 -j ACCEPT
-A INPUT -p tcp -m tcp --dport 4822 -j REJECT --reject-with tcp-reset
-A INPUT -s 127.0.0.1/32 -d 127.0.0.1/32 -p tcp -m tcp --dport 5900:5999 -j ACCEPT
-A INPUT -p tcp -m tcp --dport 5900:5999 -j REJECT --reject-with tcp-reset
-A FORWARD -d 10.1.0.100/32 -o bn1 -m set --match-set ipv4. src -j ACCEPT
-A FORWARD -d 10.1.0.100/32 -o bn1 -j ACCEPT
-A FORWARD -d 10.1.0.100/32 -o bn1 -p tcp -m tcp --dport 22 -j ACCEPT
-A FORWARD -d 10.1.0.100/32 -o bn1 -p tcp -m tcp --dport 8085 -j ACCEPT
-A FORWARD -d 10.1.0.100/32 -o bn1 -m physdev --physdev-out nbu1x1 -j LOG --log-prefix "FORWARD:TO:" --log-level 3

Reply:
i assume this 0.0.0.0/0 does not have impact on the PREROUTING chain.

Reply:
I do not see your drop rule among these. 0.0.0.0/0 matches all packets.

Reply:
@gkodali-zededa above raw table is before i applied the drop rule. my drop rule will be the end of this list.

Reply:
this is the last part of raw PREROUTING:
...
-A PREROUTING -i bn1 -m physdev --physdev-in nbu1x1+ -j LOG --log-prefix "FORWARD:FROM:" --log-level 3
-A PREROUTING -d 10.1.0.100/32 -p tcp -m tcp --dport 2375 -m physdev --physdev-in nbu1x1+ -j DROP

Reply:
sorry without the physdev stuff

Reply:
@gkodali-zededa i'll change it to the front. that's fine. i tried this:
(ns: pillar) 10d0ffa2-8df1-4c1b-be74-a8b30b0333e1:/# iptables -S -t raw
-P PREROUTING ACCEPT
-P OUTPUT ACCEPT
-A PREROUTING -d 10.1.0.100/32 -p tcp -m tcp --dport 2375 -j DROP
-A PREROUTING -i bn1 -p udp -m physdev --physdev-in nbu1x1+ -m set --match-set ipv4.local dst -m udp --dport 67 -j ACCEPT
-A PREROUTING -i bn1 -p udp -m physdev --physdev-in nbu1x1+ -m set --match-set ipv4.local src -m udp --sport 67 -j ACCEPT
-A PREROUTING -d 10.1.0.1/32 -i bn1 -p udp -m physdev --physdev-in nbu1x1+ -m udp --dport 67 -j ACCEPT
-A PREROUTING -s 10.1.0.1/32 -i bn1 -p udp -m physdev --physdev-in nbu1x1+ -m udp --sport 67 -j ACCEPT
-A PREROUTING -d 10.1.0.1/32 -i bn1 -p udp -m physdev --physdev-in nbu1x1+ -m udp --dport 53 -j ACCEPT
-A PREROUTING -s 10.1.0.1/32 -i bn1 -p udp -m physdev --physdev-in nbu1x1+ -m udp --sport 53 -j ACCEPT
-A PREROUTING -d 10.1.0.1/32 -i bn1 -p tcp -m physdev --physdev-in nbu1x1+ -m tcp --dport 53 -j ACCEPT
-A PREROUTING -s 10.1.0.1/32 -i bn1 -p tcp -m physdev --physdev-in nbu1x1+ -m tcp --sport 53 -j ACCEPT
-A PREROUTING -i bn1 -m physdev --physdev-in nbu1x1+ -m set --match-set ipv4. dst -j ACCEPT
-A PREROUTING -i bn1 -m physdev --physdev-in nbu1x1+ -j ACCEPT
-A PREROUTING -i bn1 -p tcp -m physdev --physdev-in nbu1x1+ -m tcp --sport 22 -j ACCEPT
-A PREROUTING -i bn1 -p tcp -m physdev --physdev-in nbu1x1+ -m tcp --sport 8085 -j ACCEPT
-A PREROUTING -i bn1 -m physdev --physdev-in nbu1x1+ -j LOG --log-prefix "FORWARD:FROM:" --log-level 3
(ns: pillar) 10d0ffa2-8df1-4c1b-be74-a8b30b0333e1:/#

Reply:
Thanks